tutorials/text/llama-nemotron-data-curation/main.py

```diff
@@ -0,0 +1,284 @@
+
+    # Filter out samples in thinking OFF that contain think tags
+    pipeline_thinking_off.add_stage(
+        ScoreFilter(
+            ContainsThinkOpenTagFilter(),
+            text_field="output",
+        )
+    )
+    # Filter out samples in thinking ON that do not contain think tags
+    pipeline_thinking_on.add_stage(
+        ScoreFilter(
+            MissingThinkOpenTagFilter(),
+            text_field="output",
+        )
+    )
+
+    # Filter out samples based on token count
+    tokenizer_steps = [
+        NonEnglishFilter(
+            tokenizer_identifier=args.tokenizer,
+            hf_token=args.hf_token,
+            lang_id_model_path=args.lang_id_model_path,
+            input_field="input",
+            output_field="output",
+            system_prompt_field="system_prompt",
+        ),
+        TokenCountFilter(
+            tokenizer_identifier=args.tokenizer,
+            hf_token=args.hf_token,
+            max_token_count=args.max_token_count,
+            input_field="input",
+            output_field="output",
+            system_prompt_field="system_prompt",
+        ),
+        CompletionTokenCountFilter(
+            tokenizer_identifier=args.tokenizer,
+            hf_token=args.hf_token,
+            max_completion_token_count=args.max_completion_token_count,
+            output_field="output",
+        ),
+        ApplyChatTemplate(
+            tokenizer_identifier=args.tokenizer,
+            hf_token=args.hf_token,
+            input_field="input",
+            output_field="output",
+            system_prompt_field="system_prompt",
+        ),
+    ]
+    for tokenizer_step in tokenizer_steps:
+        pipeline_thinking_on.add_stage(tokenizer_step)
+        pipeline_thinking_off.add_stage(tokenizer_step)
+
+    # No specific columns are accessed after this point, so we can drop any that the user specifies
+    if args.remove_columns:
+        # Use processing_stage decorator to remove columns
+        @processing_stage(name="remove_columns", resources=Resources(cpus=1.0), batch_size=1)
+        def remove_columns(task: DocumentBatch) -> DocumentBatch:
+            task.data = task.data.drop(columns=args.remove_columns, axis=1)
+            return task
+
+        pipeline_thinking_on.add_stage(remove_columns)
+        pipeline_thinking_off.add_stage(remove_columns)
+
+    # Save intermediate datasets
+    thinking_on_unsorted_path = os.path.join(args.output_dir, "thinking_on_unsorted")
+    thinking_off_unsorted_path = os.path.join(args.output_dir, "thinking_off_unsorted")
+    pipeline_thinking_on.add_stage(JsonlWriter(thinking_on_unsorted_path))
+    pipeline_thinking_off.add_stage(JsonlWriter(thinking_off_unsorted_path))
+
+    # Run pipelines
+    _thinking_on_output = pipeline_thinking_on.run()
+    _thinking_off_output = pipeline_thinking_off.run()
+
+    # Sort datasets
+    thinking_on_ds = ray.data.read_json(thinking_on_unsorted_path, lines=True)
```

Comment:
What's the use here? Do we need a guarantee that the output files are produced in sorted order? 
Because a global sort + write in a distributed setting, might not guarantee that. Each file would be sorted but what but if it outputs N files, they must be read in a certain way to ensure read is also sorted?
For instance spark creates files with UUID, therefore when you read the N files back there isn't a guarantee that the whole dataset in memory is sorted. That's only possible if you enforce a `coalesce` (`repartition(1)`)

Reply:
When I examined the output files, they are sorted alphabetically by name, also corresponding to the global sort operation that happened. So it worked as intended.

Toy example (not how the file names or dataset would look):
```
abc.json
   completion_token_count: 3
   completion_token_count: 7
   completion_token_count: 8

def.json
   completion_token_count: 11
   completion_token_count: 20
   completion_token_count: 90

ghi.json
   completion_token_count: 91
   completion_token_count: 92
   completion_token_count: 98
```

This is why later when I interleave the 2 sorted datasets, I use `files = sorted(f for f in os.listdir(jsonl_dir) if f.endswith(".json"))` to ensure that they are read in the correct (sorted) order.